src/org/traccar/database/DeviceManager.java

```diff
@@ -122,13 +138,16 @@ private void updateDeviceCache(boolean force) throws SQLException {
                     devicesById.remove(cachedDeviceId);
                 }
             }
-            for (String cachedDeviceUniqId : devicesByUniqueId.keySet()) {
-                if (!databaseDevicesUniqueIds.contains(cachedDeviceUniqId)) {
-                    devicesByUniqueId.remove(cachedDeviceUniqId);
+            for (String cachedDeviceUniqueId : devicesByUniqueId.keySet()) {
+                if (!databaseDevicesUniqueIds.contains(cachedDeviceUniqueId)) {
+                    devicesByUniqueId.remove(cachedDeviceUniqueId);
```

Comment:
You are right, iterator should be used here. I'll fix it
We did not get any exceptions because it is very unlikely to remove something. Only if something changed in DB directly.